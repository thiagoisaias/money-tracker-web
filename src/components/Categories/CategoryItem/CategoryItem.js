import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { devices } from "../../../utils/devices";

const Container = styled.div`
  margin: 16px 0;
  padding: 6px 0;
  border-radius: 2px;
  height: ${props => (props.isActive ? "36px" : "16px")};
  transition: ease 0.25s;
  overflow: hidden;

  &:hover {
    border-color: #ddd;
    cursor: pointer;
  }

  @media ${devices.small} {
    border-top: 1px solid #f2f2f2;
    border-bottom: 1px solid #f2f2f2;
  }

  @media ${devices.mediumUp} {
    border: 1px solid #f2f2f2;
    padding-left: 16px;
  }
`;

const FieldsContainer = styled.div`
  display: flex;
  align-items: center;
`;

const Name = styled.div`
  margin-left: 16px;
`;

const ColorField = styled.div`
  width: 1em;
  height: 1em;
  border-radius: 2px;
  background-color: ${props => props.color};
`;

const ActionsContainer = styled.div`
  display: flex;
  color: #777;
  font-size: 12px;
  margin-top: 8px;
`;

const Action = styled.div`
  margin-right: 16px;
  &:hover {
    font-weight: 600;
    cursor: pointer;
  }
`;

const CategoryItem = props => {
  const {
    categoryData,
    handleActiveItem,
    handleDelete,
    handleEdit,
    isActive
  } = props;

  return (
    <Container
      isActive={isActive}
      onClick={() => handleActiveItem(categoryData.id)}
    >
      <FieldsContainer>
        <ColorField color={categoryData.color} />
        <Name>{categoryData.name}</Name>
      </FieldsContainer>
      <ActionsContainer>
        <Action
          onClick={() => {
            handleEdit();
          }}
        >
          {"Edit"}
        </Action>
        <Action
          onClick={() => {
            handleDelete();
          }}
        >
          {"Delete"}
        </Action>
      </ActionsContainer>
    </Container>
  );
};

CategoryItem.propTypes = {
  categoryData: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired
  }).isRequired,
  handleActiveItem: PropTypes.func.isRequired,
  handleDelete: PropTypes.func,
  handleEdit: PropTypes.func,
  isActive: PropTypes.bool.isRequired
};

export default CategoryItem;
