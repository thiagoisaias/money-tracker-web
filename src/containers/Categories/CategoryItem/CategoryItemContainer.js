import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import {
  deleteCategory,
  setCategoryToEdit
} from "../../../store/actions/categories/categories";

import CategoryItem from "../../../components/Categories/CategoryItem/CategoryItem";

export class CategoryItemContainer extends Component {
  handleDelete = () => {
    const { categoryData, isLoading, onDeleteCategory } = this.props;
    if (isLoading) {
      return;
    }

    if (!window.confirm("Are you sure you want to delete this category?")) {
      return;
    }
    onDeleteCategory(categoryData.id);
  };

  handleEdit = () => {
    const { categoryData, history, onSetCategoryToEdit } = this.props;
    onSetCategoryToEdit(categoryData);
    history.push(`/categories/${categoryData.id}/edit`);
  };

  render() {
    const { categoryData, handleActiveItem, isActive } = this.props;
    return (
      <CategoryItem
        categoryData={categoryData}
        handleDelete={this.handleDelete}
        handleEdit={this.handleEdit}
        handleActiveItem={handleActiveItem}
        isActive={isActive}
      />
    );
  }
}

CategoryItemContainer.propTypes = {
  categoryData: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    color: PropTypes.string.isRequired
  }).isRequired,
  handleActiveItem: PropTypes.func.isRequired,
  isActive: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  onDeleteCategory: PropTypes.func.isRequired,
  onSetCategoryToEdit: PropTypes.func.isRequired
};

const mapStateToProps = state => {
  return {
    isLoading: state.categories.isLoading
  };
};

const mapDispatchToProps = dispatch => {
  return {
    onDeleteCategory: categoryId => {
      dispatch(deleteCategory(categoryId));
    },

    onSetCategoryToEdit: category => {
      dispatch(setCategoryToEdit(category));
    }
  };
};

export default withRouter(
  connect(mapStateToProps, mapDispatchToProps)(CategoryItemContainer)
);
