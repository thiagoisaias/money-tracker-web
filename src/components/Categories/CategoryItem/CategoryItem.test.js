import React from "react";
import ReactDOM from "react-dom";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import CategoryItem from "./CategoryItem";

describe("CategoryItem", () => {
  const mockProps = {
    categoryData: {
      id: 1,
      name: "Food",
      color: "lightblue"
    },
    handleActiveItem: jest.fn(),
    handleDelete: jest.fn(),
    handleEdit: jest.fn(),
    isActive: false
  };
  const component = shallow(<CategoryItem {...mockProps} />);

  it("renders properly", () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
