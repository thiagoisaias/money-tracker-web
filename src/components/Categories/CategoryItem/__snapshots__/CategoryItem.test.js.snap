// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryItem renders properly 1`] = `
<styled.div
  isActive={false}
  onClick={[Function]}
>
  <styled.div>
    <styled.div
      color="lightblue"
    />
    <styled.div>
      Food
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div
      onClick={[Function]}
    >
      Edit
    </styled.div>
    <styled.div
      onClick={[Function]}
    >
      Delete
    </styled.div>
  </styled.div>
</styled.div>
`;
